### PR TITLE
Reduce copying in `GetSnapshotsOperation#snapshots()`

### PR DESCRIPTION
**Pull request by DaveCTurner**
No need to create a set (the values are distinct anyway), make a
separate synchronized list, populate them both, and finally copy them
both again into another concatenated list. We can just make the final
list up front.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Reducing the unnecessary copying here is clearly good in its own right, but this change is also motivated by wanting to take more control over the `SnapshotInfo` retrieval process rather than just handing a list of snapshot IDs off to `repository.getSnapshotInfo()`. In particular if we've already found `size` matching snapshots then (depending on the chosen sort key) we might be able to tell in advance that certain other snapshots won't make it to the final list, avoiding the need to retrieve their `SnapshotInfo` blobs at all. That's the main reason for introducing a `RefCountingListener` here - it's not necessary as written yet but it will be important when we're tracking the individual retrievals ourselves.

**Comment by DaveCTurner** [comment]
https://github.com/elastic/elasticsearch/pull/105769 is what I mean by taking more control of the `SnapshotInfo` retrieval process. That PR branch includes the changes in this one so there's quite some extra noise.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 1]
```diff
@@ -400,7 +400,7 @@ private void snapshots(String repositoryName, Collection<SnapshotId> snapshotIds
             if (cancellableTask.notifyIfCancelled(listener)) {
                 return;
             }
-            final Set<SnapshotInfo> snapshotSet = new HashSet<>();
+            final List<SnapshotInfo> snapshots = new ArrayList<>(snapshotIds.size());
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:403]: >No need to create a set (the values are distinct anyway)

`snapshotIds` is defined as collection in the method signature

https://github.com/elastic/elasticsearch/blob/b95cb8c40dc701cbce18308cdcd5538fd60facb9/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java#L399

that is only called from 

https://github.com/elastic/elasticsearch/blob/b95cb8c40dc701cbce18308cdcd5538fd60facb9/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java#L378

where `toResolve`

https://github.com/elastic/elasticsearch/blob/b95cb8c40dc701cbce18308cdcd5538fd60facb9/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java#L328

is defined as a HashSet. I think it is worth to update the method signature to accept `Set<SnapshotId> snapshotIds` to better reflect that there are no duplicates expected.

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:403]: I know what you mean, but (a) nothing particularly bad happens if there _are_ duplicates here, and (b) we'd have to change that `.toList()` into a `collect(Collectors.toSet())` which is quite a bit more expensive. This whole thing will go away eventually.

**Pull request merged**